### PR TITLE
feat(validator): introduce a schema validator

### LINKED ISSUE #6: Add API schema validator

### PR DESCRIPTION
**Pull request by mvhysko**
closes #6

### CODE REVIEW

File: packages/validator/README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+short description
+
+## Setup
+
+```bash
+npm i --save @har-sdk/validator
+```
+
+## Usage
```
Comment (derevnjuk) [inline packages/validator/README.md:11]: Well, it seems there's lack of documentation)

File: packages/oas/tsconfig.json [hunk 2]
```diff
@@ -8,10 +8,13 @@
   },
   "references": [
     {
-      "path": "../openapi-sampler"
+      "path": "../openapi-sampler/tsconfig.json"
     },
     {
-      "path": "../validator"
+      "path": "../validator/tsconfig.json"
```
Comment (derevnjuk) [inline packages/oas/tsconfig.json:14]: You should specify a path to the directory which contains tsconfig.json. TSC will be able to find them by itself

**Review by derevnjuk — CHANGES_REQUESTED** [review]
Please cover by unit-tests `validator`, each edge case should be described. See details in [our KB](https://kb.neuralegion.com/#/guide/np-web-ui/scanning/discovery-types/troubleshooting?id=openapiswagger-schema-validation-checklist).

File: packages/validator/README.md (hunk 1 above)
Comment (mvhysko) [reply packages/validator/README.md:11]: added short documentation.

File: packages/validator/tests/oas.spec.ts [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+import yaml from 'js-yaml';
+import { OpenAPIV2, OpenAPIV3 } from '@har-sdk/types';
+import { expect } from 'chai';
+import { resolve } from 'path';
+import fs from 'fs';
+import { promisify } from 'util';
+
+const readFile = promisify(fs.readFile);
+
+describe('OpenApi', async () => {
```
Comment (derevnjuk) [inline packages/validator/tests/oas.spec.ts:12]: Please follow test suites/cases naming convention.

- **Unit tests**:

```ts
describe('Class name', () => {
  describe('method name', () => {
    it('should do someting', () => {
       // code
    });
  })
})
```

_Example_: see above

- **e2e/integration**:

```ts
describe('Operation name', () => {
  describe('somehow', () => {
    it('should do someting', () => {
       // code
    });
  })
})
```

_Example_:

```ts
describe('Run repeater', () => {
  describe('with --daemon option', () => {
    it('should run repeater in deamon mode', () => {
       // code
    });
  })
})
```

File: packages/validator/schemas/openapi/v3.0.0.json [hunk 4]
```diff
@@ -0,0 +1,1249 @@
+{
+  "title": "A JSON Schema for OpenAPI 3.0.",
+  "$id": "http://openapis.org/v3/schema.json#",
+  "$schema": "http://json-schema.org/draft-07/schema#",
```
Comment (derevnjuk) [inline packages/validator/schemas/openapi/v3.0.0.json:4]: Swagger 2 is based on the types supported by the draft-5 (aka Wright 00).

File: packages/validator/src/validator/DefaultValidator.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
+import { ValidatorResult, Validator } from './Validator';
+import Ajv, { ValidateFunction } from 'ajv';
+import addFormats from 'ajv-formats';
+import { Collection, getDocumentInfo, Postman } from '@har-sdk/types';
+import semver from 'semver';
+import { join, parse } from 'path';
+
+export class DefaultValidator implements Validator {
```
Comment (derevnjuk) [inline packages/validator/src/validator/DefaultValidator.ts:8]: You should extract logic to validate Postman and OAS to two separate classes. Otherwise, it will be pretty hard to maintain such spaghetti code.

File: packages/validator/src/validator/DefaultValidator.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    const schemaId = this.getSchemaId(document, info);
+    const validate: ValidateFunction | undefined = this.ajv.getSchema(schemaId);
+
+    if (!validate) {
+      throw new Error(
+        'Cannot determine version of schema. Schema ID is missed.'
+      );
+    }
+
+    if (!validate(document)) {
```
Comment (derevnjuk) [inline packages/validator/src/validator/DefaultValidator.ts:55]: You should use async version of this function

**Review by derevnjuk — CHANGES_REQUESTED** [review]

File: packages/validator/src/validator/DefaultValidator.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    3: 'http://openapis.org/v3/schema.json#'
+  };
+
+  constructor() {
+    this.ajv = new Ajv({
+      allErrors: true,
+      strict: false
+    });
+
+    addFormats(this.ajv);
```
Comment (derevnjuk) [inline packages/validator/src/validator/DefaultValidator.ts:30]: Add only these formats/keywords which are really used. That might allow us to minimize bundle size. I propose you use `rollup-plugin-visualizer` to see the difference.

**Review by derevnjuk — CHANGES_REQUESTED** [review]

File: packages/validator/package.json [hunk 8]
```diff
@@ -20,8 +20,7 @@
   "files": [
     "dist/**"
   ],
-  "main": "dist/bundle.umd.js",
-  "module": "dist/index.js",
+  "main": "dist/index.js",
```
Comment (derevnjuk) [inline packages/validator/package.json]: What is the reason to use `rollup` if you provide only es2015 module?

File: packages/validator/src/validator/OASValidator.ts [hunk 9]
```diff
@@ -0,0 +1,36 @@
+import { BaseValidator } from './BaseValidator';
+import schemaV2 from '../schemas/openapi/v2.0.0.json';
+import schemaV3 from '../schemas/openapi/v3.0.0.json';
+import { isOASV3, OpenAPI } from '@har-sdk/types';
+import semver from 'semver';
+
+export class OASValidator extends BaseValidator<OpenAPI.Document> {
+  private readonly MIN_ALLOWED_VERSION = '2.0.0';
+
+  private readonly VERSION_SCHEMA_MAP = {
```
Comment (derevnjuk) [inline packages/validator/src/validator/OASValidator.ts:10]:
```suggestion
  private readonly VERSION_SCHEMA_MAP: Readonly<Record<2 | 3, string>> = {
```

File: packages/validator/src/validator/PostmanValidator.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+import schemaV2 from '../schemas/postman/v2.0.0.json';
+import schemaV21 from '../schemas/postman/v2.1.0.json';
+import { Postman } from '@har-sdk/types';
+import { parse } from 'path';
+
+export class PostmanValidator extends BaseValidator<Postman.Document> {
+  // private readonly PATH_TO_SCHEMAS: ReadonlyArray<string> = [
+  //   'schemas/postman/v2.0.0.json',
+  //   'schemas/postman/v2.1.0.json'
+  // ];
```
Comment (derevnjuk) [inline packages/validator/src/validator/PostmanValidator.ts:11]: Remove redundant comments

File: packages/validator/src/validator/PostmanValidator.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+  constructor() {
+    super();
+    this.loadSchemas([schemaV2, schemaV21]);
+  }
+
+  protected getSchemaId(document: Postman.Document): string {
+    return document.info.schema ? parse(document.info.schema).dir + '/' : '';
```
Comment (derevnjuk) [inline packages/validator/src/validator/PostmanValidator.ts:19]: The preferred way is using template strings. BTW that's not path anyway, you can't use `path` module

File: packages/validator/src/validator/Validator.ts [hunk 12]
```diff
@@ -0,0 +1,10 @@
+import { Collection } from '@har-sdk/types';
+import { ErrorObject } from 'ajv';
+
+export interface ValidatorResult {
+  errors: Partial<ErrorObject>[];
```
Comment (derevnjuk) [inline packages/validator/src/validator/Validator.ts:5]: Extend interface by adding `status` field

File: packages/validator/tests/oas.spec.ts [hunk 13]
```diff
@@ -0,0 +1,157 @@
+import githubSwagger from './github_swagger.json';
+// import githubSwagger from './petstore_swagger.json';
```
Comment (derevnjuk) [inline packages/validator/tests/oas.spec.ts:2]: Remove useless comments

File: packages/validator/tests/oas.spec.ts [hunk 14]
```diff
@@ -0,0 +1,157 @@
+import githubSwagger from './github_swagger.json';
```
Comment (derevnjuk) [inline packages/validator/tests/oas.spec.ts:1]: Please follow naming convention, we don't use lodash in filenames

File: packages/validator/tests/oas.spec.ts [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+describe('OASValidator', async () => {
+  const validator = new OASValidator();
+
+  describe('verify', async () => {
+    it('should successfully validate GitHub swagger v2 JSON', async () => {
+      const result = await validator.verify(
+        githubSwagger as unknown as OpenAPIV2.Document
+      );
+      // eslint-disable-next-line no-console
+      console.log(result.errors);
```
Comment (derevnjuk) [inline packages/validator/tests/oas.spec.ts:22]: `console.log` is forbidden for a reason, remove it

File: packages/validator/tests/oas.spec.ts [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+import { resolve } from 'path';
+import fs from 'fs';
+import { promisify } from 'util';
+
+const readFile = promisify(fs.readFile);
+
+describe('OASValidator', async () => {
+  const validator = new OASValidator();
+
+  describe('verify', async () => {
```
Comment (derevnjuk) [inline packages/validator/tests/oas.spec.ts:16]: ditto

File: packages/validator/tests/oas.spec.ts [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+      };
+
+      try {
+        await validator.verify(apiDoc as unknown as OpenAPIV2.Document);
+      } catch (err) {
+        expect(err.message).to.be.equal(
+          'Cannot determine version of schema. Schema ID is missed.'
+        );
+      }
+    });
```
Comment (derevnjuk) [inline packages/validator/tests/oas.spec.ts:73]: This test is duplicate of previous one

File: packages/validator/tests/oas.spec.ts [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+        'utf8'
+      );
+
+      const result = await validator.verify(
+        yaml.load(content) as OpenAPIV3.Document
+      );
+      expect(result.errors.length).to.be.equal(0);
+    });
+
+    it('V2 invalid JSON: cannot determine version of document', async () => {
```
Comment (derevnjuk) [inline packages/validator/tests/oas.spec.ts:38]: Follow naming convention, https://github.com/NeuraLegion/har-sdk/pull/7#discussion_r648900275

File: packages/validator/tests/oas.spec.ts [hunk 19]
```diff
@@ -0,0 +1,157 @@
+import githubSwagger from './github_swagger.json';
+// import githubSwagger from './petstore_swagger.json';
+import { OASValidator } from '../src';
+import yaml from 'js-yaml';
+import { OpenAPIV2, OpenAPIV3 } from '@har-sdk/types';
+import { expect } from 'chai';
```
Comment (derevnjuk) [inline packages/validator/tests/oas.spec.ts:6]: use `should` instead of `expect`, see example in nexploit-service

File: packages/validator/tests/tsconfig.json [hunk 20]
```diff
@@ -0,0 +1,7 @@
+{
+  "extends": "../tsconfig.json",
+  "compilerOptions": {
+    "types": ["mocha", "node"]
+  },
+  "include": ["src", "schemas", "tests"]
```
Comment (derevnjuk) [inline packages/validator/tests/tsconfig.json:6]: Which `tests` it should include? This config is in `tests` folder) Also, why don't use `tsconfig.json` of the package itself?

File: packages/validator/src/validator/OASValidator.ts [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+    2: 'http://swagger.io/v2/schema.json#',
+    3: 'http://openapis.org/v3/schema.json#'
+  };
+
+  constructor() {
+    super();
+    this.loadSchemas([schemaV2, schemaV3]);
+  }
```
Comment (derevnjuk) [inline packages/validator/src/validator/OASValidator.ts:18]: Are really need an entire method to load schemas? Why don't pass them via constructor?
```suggestion
  constructor() {
    super([schemaV2, schemaV3]);
  }
```

**Review by derevnjuk — CHANGES_REQUESTED** [review]

File: packages/validator/src/schemas/postman/v2.0.0.json [hunk 22]
```diff
@@ -1,6 +1,7 @@
 {
   "$schema": "http://json-schema.org/draft-07/schema#",
-  "$id": "https://schema.getpostman.com/json/draft-07/collection/v2.0.0/",
+  "$id": "https://schema.getpostman.com/json/collection/v2.0.0/",
+  "$async": true,
```
[the remark this answers is not in the document]
Comment (mvhysko) [reply packages/validator/src/schemas/postman/v2.0.0.json:4]: Do you mean to put comment in schema?

File: packages/validator/src/validator/Validator.ts (hunk 12 above)
Comment (mvhysko) [reply packages/validator/src/validator/Validator.ts:5]: Do you mean this?:
```ts
enum Status {
  SUCCESS = 'success',
  FAIL = 'fail'
}
```

File: packages/validator/tests/oas.spec.ts (hunk 17 above)
Comment (mvhysko) [reply packages/validator/tests/oas.spec.ts:73]: Well, the error message is equal but the schemas are different.

File: packages/validator/src/schemas/postman/v2.0.0.json (hunk 22 above)
[the remark this answers is not in the document]
Comment (derevnjuk) [reply packages/validator/src/schemas/postman/v2.0.0.json:4]: Yeah you're right

File: packages/validator/tests/oas.spec.ts (hunk 17 above)
Comment (derevnjuk) [reply packages/validator/tests/oas.spec.ts:73]: In such way, you should adjust error message

File: packages/validator/src/validator/Validator.ts (hunk 12 above)
Comment (derevnjuk) [reply packages/validator/src/validator/Validator.ts:5]: It doesn't matter, you can add `valid` flag too

File: packages/validator/tests/postman.spec.ts [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+      try {
+        await validator.verify(apiDoc as unknown as Postman.Document);
+      } catch (err) {
+        err.message.should.be.equal(
+          'Cannot determine version of schema. Schema ID is missed.'
+        );
+      }
+    });
+
+    it('Should throw error - invalid document', async () => {
```
Comment (derevnjuk) [inline packages/validator/tests/postman.spec.ts:36]:
```suggestion
    it('should throw error if document is invalid', async () => {
```

File: packages/validator/tests/postman.spec.ts [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+
+  describe('verify', () => {
+    it('should successfully validate postman document', async () => {
+      const result = await validator.verify(
+        nexplotPostman as unknown as Postman.Document
+      );
+      result.errors.should.be.empty;
+    });
+
+    it('Should throw error - cannot determine version of document', async () => {
```
Comment (derevnjuk) [inline packages/validator/tests/postman.spec.ts:17]: ditto

File: packages/validator/tests/oas.spec.ts [hunk 25]
```diff
@@ -0,0 +1,136 @@
...
+        }
+      ];
+
+      const result = await validator.verify(
+        apiDoc as unknown as OpenAPIV3.Document
+      );
+      result.errors.should.deep.eq(errors);
+    });
+
+    it('should throw error - must have required property url', async () => {
```
Comment (derevnjuk) [inline packages/validator/tests/oas.spec.ts:108]: ditto

File: packages/validator/tests/oas.spec.ts [hunk 26]
```diff
@@ -0,0 +1,136 @@
...
+        }
+      ];
+
+      const result = await validator.verify(
+        apiDoc as unknown as OpenAPIV2.Document
+      );
+      result.errors.should.deep.eq(errors);
+    });
+
+    it('should throw error - must have required property servers', async () => {
```
Comment (derevnjuk) [inline packages/validator/tests/oas.spec.ts:81]: ditto

File: packages/validator/tests/oas.spec.ts [hunk 27]
```diff
@@ -0,0 +1,136 @@
...
+      try {
+        await validator.verify(apiDoc as unknown as OpenAPIV2.Document);
+      } catch (err) {
+        err.message.should.be.equal(
+          'Cannot determine version of schema. Schema ID is missed.'
+        );
+      }
+    });
+
+    it('should throw error - must have required property host', async () => {
```
Comment (derevnjuk) [inline packages/validator/tests/oas.spec.ts:54]: ditto

File: packages/validator/tests/oas.spec.ts [hunk 28]
```diff
@@ -0,0 +1,136 @@
...
+        'utf8'
+      );
+
+      const result = await validator.verify(
+        yaml.load(content) as OpenAPIV3.Document
+      );
+      result.errors.should.be.empty;
+    });
+
+    it('should throw error - Cannot determine version of schema. Schema ID is missed.', async () => {
```
Comment (derevnjuk) [inline packages/validator/tests/oas.spec.ts:35]: ditto

**Review by derevnjuk — APPROVED** [review]
